benchmark/common.js

```diff
@@ -88,52 +88,157 @@ Benchmark.prototype._queue = function(options) {
+    } else {
+      return result.requests.average;
+    }
+  };
 }
 
-// benchmark an http server.
-const WRK_REGEXP = /Requests\/sec:[ \t]+([0-9\.]+)/;
-Benchmark.prototype.http = function(urlPath, args, cb) {
-  hasWrk();
-  const self = this;
-
-  const urlFull = 'http://127.0.0.1:' + exports.PORT + urlPath;
-  args = args.concat(urlFull);
-
-  const childStart = process.hrtime();
-  const child = child_process.spawn('wrk', args);
-  child.stderr.pipe(process.stderr);
+function WrkBenchmarker() {
+  this.present = function() {
+    var result = child_process.spawnSync('wrk', ['-h']);
+    if (result.error && result.error.code === 'ENOENT')
+      return false;
+    else
+      return true;
+  };
+  this.create = function(path, duration, connections) {
+    const args = ['-d', duration, '-c', connections, '-t', 8,
+                  'http://127.0.0.1:' + exports.PORT + path ];
+    var child = child_process.spawn('wrk', args);
+    child.stdout.setEncoding('utf8');
+    child.stderr.pipe(process.stderr);
+    return child;
+  };
+  const regexp = /Requests\/sec:[ \t]+([0-9\.]+)/;
+  this.processResults = function(output) {
+    const match = output.match(regexp);
+    const result = match && +match[1];
+    if (!result)
+      return undefined;
+    else
+      return result;
+  };
+}
 
-  // Collect stdout
-  let stdout = '';
-  child.stdout.on('data', (chunk) => stdout += chunk.toString());
+const HTTPBenchmarkers = {
+  autocannon: new AutocannonBenchmarker(),
+  wrk: new WrkBenchmarker()
+};
 
-  child.once('close', function(code) {
-    const elapsed = process.hrtime(childStart);
-    if (cb) cb(code);
+// Benchmark an http server.
+Benchmark.prototype.http = function(urlPath, duration, connections, cb) {
+  const self = this;
+  duration = 1;
+
+  const picked_benchmarker = process.env.NODE_HTTP_BENCHMARKER ||
+                             this.config.benchmarker || 'all';
+  const benchmarkers = picked_benchmarker === 'all'
+                     ? Object.keys(HTTPBenchmarkers)
+                     : [picked_benchmarker];
+
+  // See if any benchmarker is available. Also test if all used benchmarkers
+  // are defined
+  var any_available = false;
+  for (var i = 0; i < benchmarkers.length; ++i) {
+    const benchmarker = benchmarkers[i];
+    const http_benchmarker = HTTPBenchmarkers[benchmarker];
+    if (http_benchmarker === undefined) {
+      console.error('Unknown http benchmarker: ', benchmarker);
+      process.exit(1);
+    }
+    if (http_benchmarker.present()) {
+      any_available = true;
+    }
+  }
+  if (!any_available) {
+    console.error('Couldn\'t locate any of the required http benchmarkers ' +
+                  '(' + benchmarkers.join(', ') + '). Check ' +
+                  'benchmark/README.md for further instructions.');
+    process.exit(1);
+  }
 
-    if (code) {
-      console.error('wrk failed with ' + code);
-      process.exit(code);
+  function runHttpBenchmarker(index, collected_code) {
```

Comment:
v8 optimizations is going to affect the next results if you don't restart node.js. I think it would be much better not to support this at all on the `bench.http` level, and just add an option to `bench.http` and let the configuration queue handle this. That should also reduce the code quite a bit.

```js
var bench = common.createBenchmark(main, {
  num: [1, 4, 8, 16],
  size: [1, 64, 256],
  c: [100],
  benchmarker: ['wrk', 'autocannon'] /* will run both */
});

function main(conf) {
  bench.http({
    url: '/', 
    duration: 10,
    connections: conf.c,
    benchmarker: conf.benchmarker
  }, function () { ... });
}
```